### ISSUE TITLE
Do not create Jobs and the service when JobSet is suspended

### ISSUE THREAD

**Issue by mimowo** [kubernetes-sigs/jobset#535]
To reproduce just create a simple JobSet with `spec.suspend=true`. In that case Jobs and the service are created.

This is wasteful in case of Jobs which are queued by Kueue, and may stay in the queue for a long time, potentially.

**Comment by mimowo** [comment]
Maybe not a "big" deal, but Kueue is typically used to hold long Kueues of suspended Jobs(or JobSets), say 50k, so would be nice to create them.

I imagine it would be fine to keep a once created service for a JobSet that got suspended (it was running, but got preempted). There should not be too many of them, and we could save on recreation in case the JobSet is quickly re-admitted.

**Comment by kannon92** [comment]
I think the main tricky point would be support for startup policy and suspend.

Our implementation with suspend and startup policy was to resume the replicated jobs in order of their listing.

I guess this could clean this up as we would only create the jobs if they were resumed. But it may be a bit tricky to implement...

**Comment by mimowo** [comment]
/remove-lifecycle stale
I'm investigating this so that we can fix https://github.com/kubernetes-sigs/jobset/pull/625

**Comment by kannon92** [comment]
I don't quite follow why this is required for #625. Can you explain?

**Comment by mimowo** [comment]
> I don't quite follow why this is required for #625. Can you explain?

Consider the following chain of states:  suspend - resume 1 - suspend - resume.2. 

Here resume 1 and 2 may use different Pod templates (updated during suspend), so we need to recreate the Jobs at some point. 

Deleting the in the suspend phase seems simplest. The Job controller also deletes pods on suspend

**Comment by kannon92** [comment]
I see. So Jobs delete pods on a resume/suspend but JobSet was keeping the Jobs around?

**Comment by danielvegamyhre** [comment]
>Consider the following chain of states: suspend - resume 1 - suspend - resume.2.
> Here resume 1 and 2 may use different Pod templates (updated during suspend), so we need to recreate the Jobs at some point.

@mimowo what is the current status of this issue, can we close it? We relaxed pod template mutability constraints in JobSet in order to allow for this integration with Kueue, does this not solve the issue described in your comment above?

**Comment by mimowo** [comment]
Yeah, it solves 98% (guesstimation) of problems. Let me summarize the remaining motivation to still do it (maybe as an extra option not to break the default flow):

1. transitioning JobSet via the chain (ResourceFlavor1) -> suspend -> resume (ResourceFlavor2) never remove nodeSelectors assigned before (ResourceFlavor1), just override them. This might be a problem if ResourceFlavor1 has nodeSelector: `some-key: some-value`, but ResourceFlavor2's nodeSelector does not specify the value for `some-key`. Then, it will still keep the old one in the Pod template, potentially not allowing for kubernetes-level scheduling. However, this is not an issue typically, because the ResourceFlavor2 will typically specify a new value for `some-key`.

2. For JobSets with many replicated Jobs it takes a lot of API-server resources even when the JobSet remains in Kueue.

Having said that I'm ok to close it for now, because the remaining issues haven't yet been a problem for any user I know of, so it is not a high priority to solve them proactively.

**Comment by mimowo** [comment]
IIUC it could also be helpful for lazy quota reservation for Execution Policy (see [comment](https://github.com/kubernetes-sigs/jobset/pull/680#discussion_r1789910983)), but maybe the KEP will go another way, so not sure.

**Comment by kannon92** [comment]
I have a motivation use case for this. I've been thinking about ways someone can patch or edit resources while the workload is suspended. I have heard of some use cases where people are aiming to update resources/requests on suspended resources. Some ideas are to make sure that the model can fit on the specific GPU and maybe it can be lowered if multiple GPUs are not needed. The main idea is to pick a better request/limit request based on the specification of the job. If the workload is not suspended, it would be ideal to be able to patch replicatedJobs.

See https://github.com/kubernetes/kubernetes/issues/132436.

Right now, since we create the Jobs in a suspended state, we actually are forbidden from modifying these resources due to Kubernetes restrictions. 

This could enable the above use case if we do not create the jobs.

**Comment by tchap** [comment]
I don't mind implementing this,

/assign

**Comment by GiuseppeTT** [comment]
@kannon92 @mimowo what is the current status of this request? From the discussion it seems that this is a desired change for Kueue. Is it still a problem for startup policy?

@tchap is interested in implementing it #965

**Comment by mimowo** [comment]
> @kannon92 @mimowo what is the current status of this request? 

I'm not actively working on it.

> From the discussion it seems that this is a desired change for Kueue.

Partly, currently JobSet integration has this corner-case scenario which can be solved by deleting Jobs:
- Kueue admits JobSet to RF1 and adds `nodeSelector`: "machine-type: xyz"
- Kueue preempts the JobSet (Jobs remain with the nodeSelector set)
- Kueue re-admits on RF2 which does not specify "machine-type" node selector, but only "zone: b". The nodeSelector is merged on the admission.
Issue: the Job will have both labels in the nodeSelector. 

This issue can also be solved by updating Jobs rather than re-creating. I implemented both as prototypes a while back: https://github.com/kubernetes-sigs/jobset/pull/625, https://github.com/kubernetes-sigs/jobset/pull/640. 

The remaining issue is corner case (not reported by users yet). In most cases the issue can be worked around in Kueue by specifiying some value for "machine-type" on RF2.

>  Is it still a problem for startup policy?

I'm not sure about the startup policy issue since I didn't encounter it in the prototypes. Still, I agree this is a bigger change conceptually, so one needs to be careful about it, and some API (like annotation) may be useful to drive the users choice.

**Comment by kannon92** [comment]
When I was playing around with kueue, we do not really allow users to modify resource requests to admit workloads. Say I request 8 gpus, and I realize that I can fit a queue with 4, I was thinking it would be worth having an ability to patch the existing job to admit the workload.

Yes, you can delete and recreate but maybe there are reasons to keep the underlying job.

I commented https://github.com/kubernetes-sigs/jobset/issues/535#issuecomment-2993736271 that this could be useful but I'm not so sure.

In reality, we would want to expand mutability of the JobSet and create new jobs with new resources. I am not so sure this is useful at this right now.

There is a benefit of not creating Jobs as this will be less objects in etcd. But I don't know if its worth the effort though.

WDYT @GiuseppeTT?

**Comment by tchap** [comment]
I don't know about any conceptual issues with startup policy. Actually creating and unsuspending is very similar, basically the same, so it shouldn't matter whether you create suspended jobs and unsuspend vs. create unsuspended jobs. Or at least that much I understood from the current implementation.

I would actually say the reconciliation code is easier to understand in #965

**Comment by GiuseppeTT** [comment]
> WDYT @GiuseppeTT?

Short answer: I don't think it's worth the effort either.

Long answer: I haven't had a chance to look into this issue deeply yet. My understanding is that preventing the creation of Jobs and Services when a JobSet is suspended introduces corner cases that any solution should explicitly discuss. Since this isn't blocking anyone, I would prioritize other work. However, contributions are always welcome but this particular issue requires more care.
